public/app/features/dashboard-scene/settings/VariablesEditView.test.tsx

```diff
@@ -99,6 +106,114 @@ describe('VariablesEditView', () => {
 
       errorSpy.mockRestore();
     });
+
+    it('should change the variable type creating a new variable object', () => {
+      const previousVariable = variableView.getVariables()[1] as CustomVariable;
+      variableView.onEdit('customVar2');
+
+      variableView.onTypeChange('constant');
+      expect(variableView.getVariables()).toHaveLength(2);
+      const variable = variableView.getVariables()[1] as CustomVariable;
+      expect(variable).not.toBe(previousVariable);
+      expect(variable.state.type).toBe('constant');
+
+      // Values to be kept between the old and new variable
+      expect(variable.state.name).toEqual(previousVariable.state.name);
+      expect(variable.state.label).toEqual(previousVariable.state.label);
+    });
+
+    it('should reset editing variable when going back', () => {
+      variableView.onEdit('customVar2');
+      expect(variableView.state.editIndex).toBe(1);
+
+      variableView.onGoBack();
+      expect(variableView.state.editIndex).toBeUndefined();
+    });
+
+    it('should reset editing variable when discarding changes', () => {
+      variableView.onEdit('customVar2');
+      const editIndex = variableView.state.editIndex!;
+      const variable = variableView.getVariables()[editIndex];
+      const originalState = { ...variable.state };
+
+      variable.setState({ name: 'newName' });
+      variableView.onDiscardChanges();
+
+      const newVariable = variableView.getVariables()[editIndex];
+      expect(newVariable.state).toEqual(originalState);
+    });
+
+    it('should reset editing variable when discarding changes after the type being changed', () => {
+      variableView.onEdit('customVar2');
+      const editIndex = variableView.state.editIndex!;
+      const variable = variableView.getVariables()[editIndex];
+      const originalState = { ...variable.state };
+
+      variableView.onTypeChange('constant');
+      variableView.onDiscardChanges();
+
+      const newVariable = variableView.getVariables()[editIndex];
+      expect(newVariable.state).toEqual(originalState);
+    });
+
+    it('should go back when discarding changes', () => {
+      variableView.onEdit('customVar2');
+      const editIndex = variableView.state.editIndex!;
+      expect(editIndex).toBeDefined();
+
+      variableView.onDiscardChanges();
+
+      expect(variableView.state.editIndex).toBeUndefined();
+    });
+  });
+
+  describe('Dashboard Variables dependencies', () => {
+    let variableView: VariablesEditView;
+    let dashboard: DashboardScene;
+
+    beforeEach(async () => {
+      const result = await buildTestScene();
+      variableView = result.variableView;
+      dashboard = result.dashboard;
+    });
+
+    // FIXME: This is not working because the variable is replaced or it is not resolved yet
```

Comment:
@dprokop I wrote these tests to ensure the variables keep updating the value after replacing.

The panel works, but the dashboard dependencies don't. I think I'm not doing the test well. Can you help me here?